modules/core/include/opencv2/core/dualquaternion.inl.hpp

```diff
@@ -0,0 +1,392 @@
+//                For Open Source Computer Vision Library
+//
+// Copyright (C) 2020, Huawei Technologies Co., Ltd. All rights reserved.
+// Third party copyrights are property of their respective owners.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Liangqian Kong <chargerKong@126.com>
+//         Longbu Wang <riskiest@gmail.com>
+
+#ifndef OPENCV_CORE_DUALQUATERNION_INL_HPP
+#define OPENCV_CORE_DUALQUATERNION_INL_HPP
+
+#ifndef OPENCV_CORE_DUALQUATERNION_HPP
+#error This is not a standalone header. Include dualquaternion.hpp instead.
+#endif
+
+///////////////////////////////////////////////////////////////////////////////////////
+//Implementation
+namespace cv {
+
+template <typename T>
+DualQuat<T>::DualQuat(const T vw, const T vx, const T vy, const T vz, const T _w, const T _x, const T _y, const T _z):
+                      w(vw), x(vx), y(vy), z(vz), w_(_w), x_(_x), y_(_y), z_(_z){};
+
+template <typename T>
+DualQuat<T>::DualQuat(const Vec<T, 8> &q):w(q[0]), x(q[1]), y(q[2]), z(q[3]),
+                                          w_(q[4]), x_(q[5]), y_(q[6]), z_(q[7]){};
+
+template <typename T>
+DualQuat<T> DualQuat<T>::createFromQuat(const Quat<T> &realPart, const Quat<T> &dualPart)
+{
+    T w = realPart.w;
+    T x = realPart.x;
+    T y = realPart.y;
+    T z = realPart.z;
+    T w_ = dualPart.w;
+    T x_ = dualPart.x;
+    T y_ = dualPart.y;
+    T z_ = dualPart.z;
+    return DualQuat<T>(w, x, y, z, w_, x_, y_, z_);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::createFromAngleAxisTrans(const T angle, const Vec<T, 3> &axis, const Quat<T> &trans)
+{
+    Quat<T> r = Quat<T>::createFromAngleAxis(angle, axis);
+    return createFromQuat(r, trans * r / 2);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::createFromMat(InputArray _R)
+{
+    CV_CheckTypeEQ(_R.type(), cv::traits::Type<T>::value, "");
+
+    Mat R = _R.getMat();
+    Quat<T> r = Quat<T>::createFromRotMat(R.colRange(0, 3).rowRange(0, 3));
+    Quat<T> trans(0, R.at<T>(0, 3), R.at<T>(1, 3), R.at<T>(2, 3));
+    return createFromQuat(r, trans * r / 2);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::createFromPitch(const T angle, const T d, const Quat<T> &axis, const Quat<T> &moment)
+{
+    T half_angle = angle / 2, half_d = d / 2;
+    Quat<T> qaxis = axis.normalize();
+    Quat<T> dual = Quat<T>(-half_d * std::sin(half_angle), 0.0, 0.0, 0.0) + std::sin(half_angle) * moment +
+        half_d * std::cos(half_angle) * qaxis;
```

Comment:
1. Both the moment and axis should be pure quaternions, so it's better pass them as Vec3 to avoid purity check.
2. The moment should be orthogonal to axis as said [here](https://en.wikipedia.org/wiki/Pl%C3%BCcker_coordinates), for example. This should be orthogonalized by subtracting its orthogonal part: `moment -= - qaxis*dot(qaxis, moment)`.
2. How did you get this formula? Because I have different result, and [this code](https://github.com/jhu-lcsr/handeye_calib_camodocal/blob/master/src/camodocal/calib/DualQuaternion.h#L114) has another different result too. I'm not sure which one is true one.
If you're not sure about this formula, you can use the method I used. I built quaternion from rotation and translation part where translation part is `d*qaxis + (1 - cos(angle))*cross(qaxis, m) + sin(angle)*moment)` which is quite intuitive if you draw this.

Reply:
1. OK
2. The definition of the moment is orthogonal. If the param `moment` is not, I think it's the responsibility of users. Maybe we could add the notification in the documentation?
3. The formula comes from [Dual Quaternions, Yan-Bin Jia, 2013](http://web.cs.iastate.edu/~cs577/handouts/dual-quaternion.pdf). 
In fact, this formula is the same as [this code](https://github.com/jhu-lcsr/handeye_calib_camodocal/blob/master/src/camodocal/calib/DualQuaternion.h#L114).
And now it could be written as `-half_d * std::sin(half_angle) + std::sin(half_angle) * moment +
        half_d * std::cos(half_angle) * qaxis` as I have added the adding operation for scalar and quaternion.

Reply:
1. +
2. Yes, this should be mentioned in the docs. But I also think that if we add orthogonalization, this wouldn't add too much overhead to the computations.
3. You were right, I checked my calculations - your formula is absolutely correct.

Reply:
OK